File: src/main/java/com/EmosewaPixel/pixellib/materialSystem/element/ElementUtils.java
package com.EmosewaPixel.pixellib.materialsystem.element;

import com.EmosewaPixel.pixellib.materialsystem.materials.CompoundType;
import com.EmosewaPixel.pixellib.materialsystem.materials.Material;
import com.EmosewaPixel.pixellib.materialsystem.materials.MaterialStack;
import com.EmosewaPixel.pixellib.miscutils.StreamUtils;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.DoubleStream;

//This class contains functions used for determining the elemental properties of compounds
public class ElementUtils {
    public static List<ElementStack> getElementalComposition(Material mat) {
        if (mat.getComposition().size() == 0)
            return Collections.singletonList(new ElementStack(mat.getElement()));

        Map<Element, Integer> map = mat.getFullComposition().stream().collect(Collectors.toMap(s -> s.getMaterial().getElement(), MaterialStack::getCount, StreamUtils::sum));

        return map.keySet().contains(Elements.NULL) ? Collections.singletonList(new ElementStack(Elements.NULL)) : map.entrySet().stream().map(e -> new ElementStack(e.getKey(), e.getValue())).collect(Collectors.toList());
    }

    public static int getTotalProtons(Material mat) {
        return getElementalComposition(mat).stream().mapToInt(m -> m.getElement().getProtons() * m.getCount()).sum();
    }

    public static int getTotalNeutrons(Material mat) {
        return getElementalComposition(mat).stream().mapToInt(m -> m.getElement().getNeutrons() * m.getCount()).sum();
    }

    public static double getMolarMass(Material mat) {
        return getElementalComposition(mat).stream().mapToDouble(m -> m.getElement().getAtomicMass() * m.getCount()).sum();
    }

    public static double getTotalDensity(Material mat) {
        DoubleStream stream = getElementalComposition(mat).stream().mapToDouble(m -> m.getElement().getDensity() * m.getCount());
        return mat.getCompoundType() == CompoundType.CHEMICAL ? stream.sum() : stream.average().getAsDouble();
    }
}

File: src/main/java/com/emosewapixel/pixellib/materialsystem/element/ElementUtils.kt
package com.emosewapixel.pixellib.materialsystem.element

import com.emosewapixel.pixellib.materialsystem.materials.CompoundType
import com.emosewapixel.pixellib.materialsystem.materials.Material

//This class contains functions used for determining the elemental properties of compounds
object ElementUtils {
    @JvmStatic
    fun getElementalComposition(mat: Material): List<ElementStack> {
        if (mat.composition.isEmpty())
            return listOf(ElementStack(mat.element))

        val map = mat.fullComposition.groupBy { it.material.element }.mapValues { (_, value) -> value.map { it.count }.sum() }

        return if (Elements.NULL in map) listOf(ElementStack(Elements.NULL)) else map.entries.map { ElementStack(it.key, it.value) }
    }

    @JvmStatic
    fun getTotalProtons(mat: Material) = getElementalComposition(mat).map { (element, count) -> element.protons * count }.sum()

    @JvmStatic
    fun getTotalNeutrons(mat: Material) = getElementalComposition(mat).map { (element, count) -> element.neutrons * count }.sum()

    @JvmStatic
    fun getMolarMass(mat: Material) = getElementalComposition(mat).map { (element, count) -> element.atomicMass * count }.sum()

    @JvmStatic
    fun getTotalDensity(mat: Material): Double {
        val stream = getElementalComposition(mat).map { (element, count) -> element.density * count }
        return if (mat.compoundType === CompoundType.CHEMICAL) stream.sum() else stream.average()
    }
}